core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,62 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+case class Perm(mapping: Map[Int, Int]) extends AnyVal {
+  def apply(k: Int): Int = mapping.getOrElse(k, k)
+  def image: Set[Int] = mapping.keySet
```

Comment:
Technically, the image is also the preimage, as assumed by line [53 below](https://github.com/typelevel/spire/pull/794/files#diff-f10faa3e5d33a83c27a54a22b3bed65eR53).

I could add an assertion that `mapping.keySet == mapping.values.toSet`